src/main/java/org/myrobotlab/service/Runtime.java

```diff
@@ -1144,145 +1128,98 @@ static public boolean loadAll() {
-            repo.install(services.get(i));
+          for (String service : options.install) {
+            repo.install(service);
           }
         }
         shutdown();
         return;
       }
 
-      if (cmdline.containsKey("--manifest") || cmdline.containsKey("-m")) {
+      if (options.manifest) {
         CodecJson.encode(Runtime.getManifest());
         shutdown();
         return;
       }
 
-      if (cmdline.containsKey("--extract") || cmdline.containsKey("-e")) {
-        // force all updates
-        /*
-         * FIXME - do
-         * " -extract {serviceType} in future ArrayList<String> services = cmdline.getArgumentList("
-         * -extract"); Repo repo = Repo.getInstance(); if (services.size() == 0)
-         * { repo.install(); } else { for (int i = 0; i < services.size(); ++i)
-         * { repo.install(services.get(i)); } }
-         */
+      if (options.extract) {
         extract();
         shutdown();
         return;
       }
 
-      if (cmdline.containsKey("-service")) {
-        createAndStartServices(cmdline);
+      if (options.services != null) {
+        createAndStartServices(options.services);
       }
 
-      if (cmdline.containsKey("-invoke")) {
-        invokeCommands(cmdline);
+      if (options.invoke != null) {
+        invokeCommands(options.invoke);
       }
 
     } catch (Exception e) {
-      Logging.logError(e);
-      System.out.print(Logging.stackToString(e));
-      Service.sleep(2000);
+      Runtime.mainHelp();
+      shutdown();
+      log.error("main threw", e);
     }
   }
 
   /**
    * prints help to the console
    */
   static void mainHelp() {
-    System.out.println(String.format("Runtime %s", Runtime.getVersion()));
-    System.out.println("-h --help                                  # help ");
-    System.out.println("-v --version                               # print version");
-    System.out.println("-update                                    # update myrobotlab");
-    System.out.println("-invoke name method [param1 param2 ...]    # invoke a method of a service");
-    System.out.println("-install [ServiceType1 ServiceType2 ...]   # install services - if no ServiceTypes are specified install all");
-    System.out.println("-id <instance id>                          # the identifier for this process");
-    System.out.println("-logToConsole                              # redirects logging to console");
-    System.out.println("-logLevel <DEBUG | INFO | WARNING | ERROR> # log level");
-    System.out.println("-service <name1 Type1 name2 Type2 ...>     # create and start list of services, e.g. -service gui SwingGui");
-    // System.out.println("example:");
-    // String helpString = "java
-    // -Djava.library.path=./libraries/native/x86.32.windows
-    // org.myrobotlab.service.Runtime -service webgui WebGui gui SwingGui
-    // -logLevel INFO -logToConsole";
-    // System.out.println(helpString);
+    new CommandLine(new CmdOptions()).usage(System.out);

```

Comment:
nice.

Reply:
yes a lovely piece of software - picocli